Adding the state-of-the-art contrastive search decoding methods for the codebase of generation_utils.py

gmftbyGMFTBY opened the pull request:
# Adding the state-of-the-art contrastive search decoding method for the `generation_utils` codebase

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #19182

In this PR, I add the source codes of our proposed state-of-the-art decoding methods for the off-the-shelf neural text generation models. The main changes are in the following files: (1) `src/transformers/generation_utils.py`; (2) `examples/pytorch/text-generation/run_generation_contrastive_search.py`. To run the test script, please follow these commands:

```bash
cd examples/pytorch/text-generation;
CUDA_VISIBLE_DEVICES=0 python run_generation_contrastive_search.py --model_type=gpt2 --model_name_or_path=gpt2-large
```

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests), Pull Request section?
- [x] The PR has been well discussed in [19182](https://github.com/huggingface/transformers/issues/19182)
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests? Yes, I have written the test scripts for the [contrastive search](https://github.com/gmftbyGMFTBY/transformers/blob/csearch-pr-v2/examples/pytorch/text-generation/run_generation_contrastive_search.py)


## Who can review?

According to the suggestions of @gante, @patrickvonplaten and @sgugger can review this PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

gmftbyGMFTBY committed:
add: the contrastive search for generaton_utils

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
add: testing scripts for contrastive search under examples/text-generation

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_19477). All of your documentation changes will be reflected on that endpoint.

gmftbyGMFTBY committed:
update the quality of codes

gmftbyGMFTBY committed:
Merge branch 'csearch-pr-v2' of https://github.com/gmftbyGMFTBY/transformers into csearch-pr-v2

gante commented:
@sgugger @patrickvonplaten context: this is the implementation by the authors of [this NeurIPS paper](https://arxiv.org/abs/2202.06417) -- a new generation strategy with very interesting results!

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -1634,6 +1731,229 @@ def typeerror():
+        **model_kwargs,
+    ) -> Union[GreedySearchOutput, torch.LongTensor]:
+        r"""
+        Parameters:
+        Generates sequences of token ids for models with a language modeling head using **contrastive search** and can
+        be used for text-decoder, text-to-text, speech-to-text, and vision-to-text models.
```
```suggestion
        Generates sequences of token ids for models with a language modeling head using **contrastive search** and can
        be used for text-decoder, text-to-text, speech-to-text, and vision-to-text models.
        
        Parameters:
```
The parameters block goes after the introduction.

sgugger commented on src/transformers/generation_utils.py:
````diff
@@ -1634,6 +1731,229 @@ def typeerror():
+            model_kwargs:
+                Additional model specific keyword arguments will be forwarded to the `forward` function of the model.
+                If model is an encoder-decoder model the kwargs should include `encoder_outputs`.
+        Return:
+            [`~generation_utils.GreedySearchDecoderOnlyOutput`], [`~generation_utils.GreedySearchEncoderDecoderOutput`]
+            or `torch.LongTensor`: A `torch.LongTensor` containing the generated tokens (default behaviour) or a
+            [`~generation_utils.GreedySearchDecoderOnlyOutput`] if `model.config.is_encoder_decoder=False` and
+            `return_dict_in_generate=True` or a [`~generation_utils.GreedySearchEncoderDecoderOutput`] if
+            `model.config.is_encoder_decoder=True`.
+        Examples:
+        ```python
+        >>> from transformers import (
+        ...     AutoTokenizer,
+        ...     AutoModelForCausalLM,
+        ...     LogitsProcessorList,
+        ...     MinLengthLogitsProcessor,
+        ...     StoppingCriteriaList,
+        ...     MaxLengthCriteria,
+        ... )
````
```suggestion

        Return:
            [`~generation_utils.GreedySearchDecoderOnlyOutput`], [`~generation_utils.GreedySearchEncoderDecoderOutput`]
            or `torch.LongTensor`: A `torch.LongTensor` containing the generated tokens (default behaviour) or a
            [`~generation_utils.GreedySearchDecoderOnlyOutput`] if `model.config.is_encoder_decoder=False` and
            `return_dict_in_generate=True` or a [`~generation_utils.GreedySearchEncoderDecoderOutput`] if
            `model.config.is_encoder_decoder=True`.

        Examples:
        ```python
        >>> from transformers import (
        ...     AutoTokenizer,
        ...     AutoModelForCausalLM,
        ...     LogitsProcessorList,
        ...     MinLengthLogitsProcessor,
        ...     StoppingCriteriaList,
        ...     MaxLengthCriteria,
        ... )
```
Leave room for readability.

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -3446,3 +3766,152 @@ def top_k_top_p_filtering(
     return logits
+
+
+# ========== utils for contrastive search decoding method ========= #
```
Those utils would probably be best in a submodule (like we have `generation_beam_search` or `generation_beam_constraint`) to avoid `generation_utils` being too big.

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -921,6 +970,7 @@ def generate(
         early_stopping: Optional[bool] = None,
         num_beams: Optional[int] = None,
         temperature: Optional[float] = None,
+        penalty_alpha: Optional[float] = None,
```
This new arg should be documented in the docstring.

sgugger commented on examples/pytorch/text-generation/run_generation_contrastive_search.py:
```diff
@@ -0,0 +1,278 @@
+                language = input("Using XLM. Select language in " + str(list(available_languages)) + " >>> ")
+
+        model.config.lang_id = model.config.lang2id[language]
+        # kwargs["language"] = tokenizer.lang2id[language]
+
+    # TODO fix mask_token_id setup when configurations will be synchronized between models and tokenizers
+    # XLM masked-language modeling (MLM) models need masked token
+    # is_xlm_mlm = "mlm" in args.model_name_or_path
+    # if is_xlm_mlm:
+    #     kwargs["mask_token_id"] = tokenizer.mask_token_id
```
To clean up?

sgugger reviewed (commented):
Thanks a lot for adding this new method. The added code in `generate` looks clean to me, I just left a couple of nits.

For the example, it would really be good if you could write a new one leveraging the auto-APIs instead of copying the old `run_generation.py` which is severly outdated and only works for very few models.

gmftbyGMFTBY commented:
Hi, @sgugger, thank you so much for your suggestions. I will fix these problems quickly!

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY commented:
Hello @sgugger, is there any document or introduction for auto-APIs?

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
revise the docstring; make the generation_contrastive_search.py scripts;

sgugger commented:
The [documentation](https://huggingface.co/docs/transformers/model_doc/auto) would be the place to start. You can also look at all other examples!

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
revise the examples/pytorch/text-generation/run_generation_contrastive_search.py to the auto-APIs format

gmftbyGMFTBY commented:
Hello, @sgugger, I have fixed the problems based on your valuable suggestions! Besides, I have updated the test scripts to the auto-APIs of inference.

The command line to run this test script can be found in its docstring.

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
revise the necessary documents

sgugger reviewed (commented):
Mostly have nits on the docstrings. The new example looks great, thanks a lot!

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
fix: revise the docstring of generation_contrastive_search.py

gmftbyGMFTBY committed:
Fix the code indentation

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
fix: revise the nits and examples in contrastive_search docstring.

gmftbyGMFTBY commented:
Hello @sgugger, I have fixed the problems based on your suggestions.

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gante commented on examples/pytorch/text-generation/run_generation_contrastive_search.py:
```diff
@@ -0,0 +1,139 @@
+# limitations under the License.
+""" The examples of running contrastive search on the auto-APIs;
+
+Running this examples:
```
```suggestion
Running this example:
```

gante commented on examples/pytorch/text-generation/run_generation_contrastive_search.py:
```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python
+# coding=utf-8
+# Copyright 2018 Google AI, Google Brain and Carnegie Mellon University Authors and the HuggingFace Inc. team.
+# Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
```
You should update this to contain this year (2022) and your affiliation :)

gante commented on examples/pytorch/text-generation/run_generation_contrastive_search.py:
```diff
@@ -0,0 +1,139 @@
+""" The examples of running contrastive search on the auto-APIs;
+
+Running this examples:
+CUDA_VISIBLE_DEVICES=0 python run_generation_contrastive_search.py --model_name_or_path=gpt2-large --penalty_alpha=0.6 --k=4 --length=256
```
Is `CUDA_VISIBLE_DEVICES=0` required to reproduce the results?

If no, I'd remove it (and the user will add it if needed). If yes, then we should understand why

gmftbyGMFTBY replied:
The `CUDA_VISIBLE_DEVICES` parameter is used for speeding up the inference, which doesn't affect the results. I think I should remove it.

gante commented on src/transformers/generation_contrastive_search.py:
```diff
@@ -0,0 +1,199 @@
+# coding=utf-8
+# Copyright 2020 The Google AI Language Team Authors, Facebook AI Research authors and The HuggingFace Inc. team.
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
This also needs to be updated (as above)

gante commented on src/transformers/generation_contrastive_search.py:
```diff
@@ -0,0 +1,199 @@
+logger = logging.get_logger(__name__)
+
+
+"""
+This file contains the utils functions for the contrastive search, which will be called in `generation_utils`
+"""
```
This should be moved to right after the license, before the imports

gante reviewed (approved):
LGTM 👍  (added a few minor nits)

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
fix the copyright

patrickvonplaten commented on src/transformers/__init__.py:
```diff
@@ -3750,6 +3751,7 @@
             PhrasalConstraint,
         )
         from .generation_beam_search import BeamScorer, BeamSearchScorer, ConstrainedBeamSearchScorer
+        from .generation_contrastive_search import ContrastiveDecodingOneStepFast
```
I would maybe not put this in the public init as it could be prone to change in the future and I don't think most people will use it outside of `generate` no?

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1410,6 +1493,27 @@ def generate(
+
+            if num_return_sequences > 1:
+                raise ValueError(
+                    f"num_return_sequences has to be 1, but is {num_return_sequences} when doing greedy search."
```
```suggestion
                    f"num_return_sequences has to be 1, but is {num_return_sequences} when doing contrastive search."
```

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1645,6 +1749,237 @@ def typeerror():
+
+        # encode the given prefix and prepare model inputs; encoder-decoder model process the prefix and save the `encoder_outputs`
+        model_kwargs["use_cache"] = True
+        model_kwargs["past_key_values"] = None
```
Why do we set this to `None` here ? It should never be not None IMO as no forward pass has been done yet. Can't we remove this line?

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1645,6 +1749,237 @@ def typeerror():
+        this_peer_finished = False  # used by synced_gpus only
+
+        # encode the given prefix and prepare model inputs; encoder-decoder model process the prefix and save the `encoder_outputs`
+        model_kwargs["use_cache"] = True
```
Why is this necessary? One always has to make use of the cache?

gmftbyGMFTBY replied:
Since contrastive search needs to forward twice in a step to select the next token, the inference speed is slightly slower than other decoding methods. However, this problem can be easily solved by using the cache computed in the previous steps. Thus, IMO the contrastive search should use the cache mechanism during decoding.

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1645,6 +1749,237 @@ def typeerror():
+                else:
+                    this_peer_finished = True
+
+            # prepare model inputs
```
Not a huge fan of having this down here as it's different from the logic of the other generate methods. Could we somehow move it up? Potentially skipping the very first step as it's already done? At the moment this will always be run once unnecessarily at the end

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1645,6 +1749,237 @@ def typeerror():
+        # encode the given prefix and prepare model inputs; encoder-decoder model process the prefix and save the `encoder_outputs`
+        model_kwargs["use_cache"] = True
+        model_kwargs["past_key_values"] = None
+        model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
```
is there any way we could move this inside the loop? Think this would be cleaner. Maybe we can do something like:


```python

output_hidden_states = last_hidden_states is None
...
output = self(....)
...
if last_hidden_states is None:
    last_hidden_states = output.hidden_states[-1]

```

patrickvonplaten commented on src/transformers/generation_contrastive_search.py:
```diff
@@ -0,0 +1,196 @@
+
+
+def ContrastiveDecodingOneStepFast(
+    model,
```
not very happy about passing the `model` into the function here and upper-casing the function.

Could we try to have this whole functionality directly inside the `contrastive_search` function? It would help readability a lot IMO.

@gante - what do you think?

patrickvonplaten commented on src/transformers/generation_contrastive_search.py:
```diff
@@ -0,0 +1,196 @@
+
+    # compute the degeneratin penalty and re-rank the candidates based on the degeneration penalty and the model confidence
+    # the scores and index of the selected tokens are returned
+    selected_scores, selected_idx = ranking_fast(
```
@gante I think we can/should make this a logit processor that exceptionally takes `context_hidden` and `next_hidden` as inputs arguments as well . `top_k_probs` can be computed inside the logit processor, `penalty_alpha` and `beam_width` can be stored inside the logic processor at init

patrickvonplaten commented on src/transformers/generation_contrastive_search.py:
```diff
@@ -0,0 +1,196 @@
+logger = logging.get_logger(__name__)
+
+
+def ranking_fast(
```
If possible I'd be in favor of not adding a new file here. IMO `ranking_fast` can be changed to a logit processor and `ContrastiveDecodingOneStepFast` is could be fully moved into `generation_utils.py` right away. 

IMO this will be easier to maintain and understand for people that know already how `generate` works - wdyt @gante ?

patrickvonplaten commented:
Hey @gmftbyGMFTBY,

Super nice PR! Contrastive search decoding would be a great addition to Transformers. I feel a bit uneasy of the logic in src/transformers/generation_contrastive_search.py and would prefer to not add a new file here and instead try to adapt the code more to our existing design (try to make use of logits processor, call the model **only** in the loop and not before).

We usually haven't added any new files for generation and molding the design more into what already exists would help a lot with maintainability. 
As an example for constrained decoding, we sadly don't manage to maintain the function anymore (see https://github.com/huggingface/transformers/pull/17920) because the code was too different & too hard for us to maintain (IMO).

If possible I'd advocate quite strongly for trying to make the PR a bit more compatible with our current design (it shouldn't require much work IMO).

@gante @sgugger what do you think?

sgugger commented:
I don't know the generate code well enough to evaluate how easy/hard it would be for this PR to fit in the existing design. I solely gave my approval based on the fact it was mimicking something that existed with constraint decoding. Sorry, I wasn't aware it wasn't maintained anymore.

Will know for the next time :-)

gante commented:
@patrickvonplaten replying here since I believe the multiple important points you raised are related. This will be a bit long, but bear with me - I think we can improve the quality of the PR before it gets merged 🙌 

1. `ContrastiveDecodingOneStepFast` (which should be snake cased, perhaps into `contrastive_decoding_step`) being a stand-alone function - it could be part of the loop, yes, but this is a long operation that in essence replaces picking the argmax (in `greedy_search`) or sampling (in `sample`). I agree it should not be public, but I believe we would stand to gain from a readability standpoint if we separate it somehow -- perhaps a private method in the MixIn or an inner function to `contrastive_search`? These two options would also mean that we don't need to pass `model`, which is indeed awkward.
2. `ranking_fast` being a `LogitsProcessor` - I would disagree here, despite being an operation that processes logits. In addition to breaking the fairly stable `call()` API with two new tensor inputs (`context_hidden` and `next_hidden`), it can only be used with `contrastive_search`. As such, I think we would only be adding a confusing `LogitsProcessor` to the public API.
3. Despite the above, I agree that we should pass and compute the logits processors before the decoding function, like the other `generate` methods -- this was an oversight on my end at review time! As for `top_k`, I have no strong feelings - on one hand, `contrastive search` is not meant to be used with other `LogitWarpers` (like `TemperatureLogitsWarper`), on the other hand there is probably no harm in using `LogitWarpers`.

(now the most challenging issue in design, IMO)

4. A naive implementation of `contrastive_search` needs two sets of forward passes per token -- once to get the `top_k` candidate tokens, another to get the future hidden states for each candidate (to be used as a filter). We can avoid this 2x cost if we pipe future `past_key_values` corresponding to the selected candidate into the actual `past_key_values`. It does mean, however, that the first token requires an additional forward pass. The current code does it by placing `prepare_inputs_for_generation()` (and a few other ops) in unconventional places, before the loop and at the end of the loop. Perhaps it would be better if we kept the original structure, but added an `if` at the start loop -- if we are in the first iteration, do this set of operations.

WDYT about these 4 points?

patrickvonplaten commented:
Thanks for the summary @gante !

Regarding the 4 points above:

1.) I think the `ContrastiveDecodingOneStepFast` is the core of the contrastive generation algorithm and I don't see any advantage having it it's own "sub-function" (it'll never be called from another function than `contrastive_search`).  Also, at the moment, the method returns 6 tensors and accepts 8 or so arguments so it's much more than just replacing the greedy argmax operation IMO. To begin with, I think it'd be very helpful to just copy-paste all the code into the `contrastive_search` and then later down the road I think we could see if parts of it could be moved out (don't think it's necessary though).

2.) I think we would have the following gains to having it implemented as a logit processor:
- We understand the method better & easier to maintain. If we mold `ranking` fast into a logit processor, it's much easier to understand it
- Very easy to test this method
- It is to me a method that processes logits -> so logically it should be a logits processor to me
- Now regarding the API, IMO `logits_processor` inputs are not just restricted to `input_ids` and `scores`. IMO, the API is rather (`input_ids`, `scores`, <other-args-that-are-needed>) => `scores` . Also, we've already "broken" the "input_ids" & "scores" - only API here: https://github.com/huggingface/transformers/blob/d2e5b19b821f0cf43c7cf4f01be5faa1cb20aa64/src/transformers/generation_utils.py#L2989 
=> I understand your point here and ok for me to not make it a logits processor, but IMO it would be better/cleaner
3.) Yes I think it's very important to align the structure of the generate method as much as possible to the other ones
4.) Totally fine for me to have an extra forward pass in the beginning, it's just important that we make the structure the same as greedy or sample (I don't have a problem with if statements here at all). For maintenance and also to understand the method, it's very important IMO that one could compare greedy search and constrastive search line-by-line and then see quickly how the two methods differ

gmftbyGMFTBY commented:
Thank you for your valuable responses! I will revise this PR according to your reviews.

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gante commented:
@gmftbyGMFTBY let's go with the core of @patrickvonplaten's suggestions. Patrick and I also talked on Slack to align a few details regarding `ranking_fast` :D

Here's a summary of the main changes we are requesting, to ensure the code you're adding remains easy to maintain:
1. Move the contents of `ContrastiveDecodingOneStepFast` to where the function is called, as opposed to being a function call;
2. Keep `ranking_fast` as it is now (i.e. NOT a logits processor);
3. Let's apply the logits processors at the start of each iteration, and move `top_k` to the logits warpers (like `sample` [does it](https://github.com/huggingface/transformers/blob/d2e5b19b821f0cf43c7cf4f01be5faa1cb20aa64/src/transformers/generation_utils.py#L2059)). This ensures that there are minimal differences between generation strategies;
4. Let's rearrange the order of operations such that all model forward passes happen inside the generation loop (with an `if` for the operations that are only supposed to happen on the first iterations).

Here is a diagram with the expected code structure, to ensure we are all on the same page: https://miro.com/app/board/uXjVPMiVFFg=/

Finally, thank you for your cooperation with us 🙌 This back and forth may be a bit frustrating, but it will ensure your contribution will be long lived!

gmftbyGMFTBY commented:
@gante Nice! but I just saw your message after I finished writing the `logits_processor` corresponding to `ranking_fast` function.
The followings are the implementation of the `ranking_fast`'s `logits_processor`:
![image](https://user-images.githubusercontent.com/27548710/196190366-5cb9c697-051d-49de-903d-b4279b3d8571.png)

It is initialized in `_get_logits_processor` function:
![image](https://user-images.githubusercontent.com/27548710/196190683-70561c32-c38a-4991-bc01-e058b3378a34.png)

and will be called:
![image](https://user-images.githubusercontent.com/27548710/196190797-fb7568d1-99d2-400d-a45b-205522b8309f.png)

Can I keep this implementation or just still employ the `ranking_fast` function?

gmftbyGMFTBY commented:
@gante Oh, I got your reason about not using the `logits_processor`. I will follow the instruction in the `sample` function!

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
delete generation_contrastive_search.py

gmftbyGMFTBY committed:
Merge branch 'csearch-pr-v2' of https://github.com/gmftbyGMFTBY/transformers into csearch-pr-v2

update the codes with transformers

gmftbyGMFTBY commented:
The revisions have been updated!

gante commented on src/transformers/generation_utils.py:
```diff
@@ -1645,6 +1759,330 @@ def typeerror():
+            # contrastive search decoding consists of two steps: (1) candidate tokens recall; (2) candidate re-rank by degeneration penalty
+            bsz, seqlen, embed_dim = last_hidden_states.size()
+
+            # logits processor: empty logits processor
```
```suggestion
```
This comment is not necessarily true -- e.g. if the user passes `bad_words_ids` to `generate()` then `NoBadWordsLogitsProcessor` will be in `logits_processor`

gante commented on src/transformers/generation_utils.py:
```diff
@@ -1645,6 +1759,330 @@ def typeerror():
+
+            # logits processor: empty logits processor
+            logit_for_next_step = logits_processor(input_ids, logit_for_next_step)
+            next_probs = nn.functional.softmax(logit_for_next_step, dim=-1)
+            # logit warper: applying after the softmax, which is consistent with the logic in the paper
+            logit_for_next_step = logits_warper(input_ids, next_probs)
+
+            _, top_k_ids = torch.topk(logit_for_next_step, dim=-1, k=top_k)
+            top_k_probs = torch.gather(next_probs, dim=1, index=top_k_ids)
```
Uhmmm I see. Because the top k IDs are needed, and our `logits_warper` only returns the updated scores, a `torch.topk()` call outside the logits warper is always needed 🤔 The `logits_warper` already performs a `top_k` call, but it only filters the scores. However, `logits_warper` may also include other operations (e.g. nucleus sampling-like filtering, with `top_p`), so I think it's positive to be included here.

A potential change might be needed, though: `logits_warper` can change the logits, so the softmax should happen after that line.

gante commented on src/transformers/generation_utils.py:
```diff
@@ -1645,6 +1759,330 @@ def typeerror():
+            if step_counter == 0:
+                # encode the given prefix and prepare model inputs; encoder-decoder model process the prefix and save the `encoder_outputs`
+                model_kwargs["use_cache"] = True
+                model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
```
hah, it seems my diagram was slightly wrong! This should be out of the loop, and the same call below should be removed

👉 https://miro.com/app/board/uXjVPMiVFFg=/

gante commented on src/transformers/generation_utils.py:
```diff
@@ -1645,6 +1759,330 @@ def typeerror():
+
+            # prepare model inputs
+            model_kwargs["past_key_values"] = past_key_values
+            model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
```
```suggestion
```

gante commented on src/transformers/generation_utils.py:
```diff
@@ -3456,3 +3894,25 @@ def top_k_top_p_filtering(
+    norm_context_hidden = context_hidden / context_hidden.norm(dim=2, keepdim=True)
+    norm_next_hidden = next_hidden / next_hidden.norm(dim=2, keepdim=True)
+    cosine_matrix = torch.matmul(norm_context_hidden, norm_next_hidden.transpose(1, 2)).squeeze(-1)  # [B*K, S]
+    scores, _ = torch.max(cosine_matrix, dim=-1)  # [B*K]
+    next_top_k_probs = next_top_k_probs.view(-1)  # [B*K]
+    scores = (1.0 - alpha) * next_top_k_probs - alpha * scores
+    scores = torch.stack(torch.split(scores, beam_width))  # [B, K]
+    selected_scores, selected_idx = scores.max(dim=-1)  # [B]
+    return selected_scores, selected_idx
```
⚠️ our `scores` are always in the log domain, so I suggest two changes:
1. rename `scores` into something else to prevent mixing them up (e.g. `contrastive_score` would be different enough)
2. return `torch.log(selected_scores)`

gante reviewed (commented):
A few additional (minor) suggestions and requests :)

I think the PR is close to a ready state. The last significant step missing is some form of test, to ensure we don't regress. An integration test is enough for now. You can add them [here](https://github.com/huggingface/transformers/blob/c7edde1a692012eda23bc2b837588557b97ad729/tests/generation/test_generation_utils.py#L1661), and as you can see in the examples, it's pretty straightforward -- simply test that we obtain the expected output for a given input. I suggest you use a test case more complex than your docstring example (e.g. a longer output with a slightly bigger model, preferably an example from your paper)

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
revise the logic in contrastive_search

gmftbyGMFTBY commented:
Hello, @gante, I have updated the PR based on your suggestions.

gmftbyGMFTBY commented:
Oh, I am still working on the integration test.

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
update the intergration test and the docstring

gmftbyGMFTBY committed:
run the tests over

gante commented:
@gmftbyGMFTBY you probably have to add the `@slow` decorator to the test, and run it with `RUN_SLOW=1 py.test (...)`. We don't run tests with `@slow` on push, but we run it every 24h and track it internally :)

gmftbyGMFTBY commented:
Ok, I got it!

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
add the slow decorate to the contrastive_search intergrate test

gante commented on src/transformers/generation_utils.py:
```diff
@@ -1500,22 +1500,11 @@ def generate(
-                temperature=temperature,
-                num_beams=num_beams,
-                renormalize_logits=renormalize_logits,
-            )
```
This shouldn't have been removed :)

gmftbyGMFTBY replied:
During the test, I found that adding the logits_warper (the TopKLogitsWarper is used) could influence the generations of contrastive search. Because the TopKLogitsWarper filters the logits of other tokens (not Top-k tokens) and calculates the `softmax`, the `model confidence` is different from the case that the TopKLogitsWarper is not used. 

So, in this case, I think the contrastive search should disable the logits_warper by default. 

Is there any solution that the TopKLogitsWarper warper is not activated for contrastive search?

gante replied:
Interesting. In that case, I think we can do without the logits warper for now :)

gmftbyGMFTBY replied:
Okay!

gante commented on src/transformers/generation_utils.py:
```diff
@@ -1500,22 +1500,11 @@ def generate(
                 top_k=top_k,
                 penalty_alpha=penalty_alpha,
                 logits_processor=logits_processor,
-                logits_warper=logits_warper,
```
(related to the comment above)

gante reviewed (approved):
Cool, seems ready to go in (except for the minor comment I added)

@patrickvonplaten can you do a final check, please? :)

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1646,6 +1749,318 @@ def typeerror():
+        logits_processor: Optional[LogitsProcessorList] = None,
+        logits_warper: Optional[LogitsProcessorList] = None,
+        stopping_criteria: Optional[StoppingCriteriaList] = None,
+        max_length: Optional[int] = None,
```
```suggestion
```

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1646,6 +1749,318 @@ def typeerror():
+            stopping_criteria (`StoppingCriteriaList`, *optional*):
+                An instance of [`StoppingCriteriaList`]. List of instances of class derived from [`StoppingCriteria`]
+                used to tell if the generation loop should stop.
+            max_length (`int`, *optional*, defaults to 20):
```
```suggestion
```

patrickvonplaten replied:
Let's maybe not add a deprecated argument :-)

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1646,6 +1749,318 @@ def typeerror():
+            )
+            # compute the candidate tokens by the language model and collects their hidden_states
+            output = self(output_hidden_states=True, **next_model_inputs)
+            past_key_values = output.past_key_values
```
(nit) - not all language gerenation models can return `past_key_values` (e.g. TransfoXL or XLNet) these models are still surprisingly used a lot: 
- https://huggingface.co/xlnet-base-cased (> 200k downloads)

Maybe we could add a better error message here? 
```suggestion
            if "past_key_values" not in output:
                raise ValueError(f"self.__class__ cannot return `past_key_values` and can therefore **not** be used for contrastive generation.")
            past_key_values = output.past_key_values
```

gmftbyGMFTBY replied:
Okay!

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1646,6 +1749,318 @@ def typeerror():
+                items = []
+                # item is either the key or the value matrix
+                for item in layer:
+                    bsz_and_beam, num_head, seq_len, esz = item.size()
```
I'm not sure it holds always true that the `past_key_values` have this size. Did we test contrastive search on all of the following models:

- GPT2
- T5
- GPT-J
- BART
?

It should work at least on those 4 models

gmftbyGMFTBY replied:
Yes, GPT2, T5, BART, GPT-J, and OPT models work fine.

Our implementation is compatible with the encoder-decoder models (the degeneration penalty is calculated on the decoder's hidden states). But we didn't carefully conduct the human evaluation of the encoder-decoder models, such as T5 and BART. Whether contrastive search could significantly boost their performance is still an open problem for us.

patrickvonplaten commented on tests/generation/test_generation_utils.py:
```diff
@@ -1693,6 +1693,25 @@ def test_diverse_beam_search(self):
         )
 
+    @slow
+    def test_contrastive_search(self):
```
If possible I'd be really happy if we could also test this on BART, T5 and GPT-J . Then we should have covered 95% of the model architectures. But ok to do in a follow-up PR . Currently I don't expect the method to work with T5.

patrickvonplaten reviewed (approved):
Ok for me to merge, but it'd be nice to always make sure the method works for all T5, BART, GPT2 and GPT-J. Also, we currently have only slow tests which is dangerous given that changes in generate can also affect contrastive search.

If it doesn't take too much time, I'd advocate to at least add 7 more tests:
- 4 fast tests with dummy models that just check constrastive search outputs the correct shape, one for each GPT2, T5, BART, GPT-J
- 3 more slow tests exactly like `test_contrastive_search` for T5, BART, GPT-J

I leave it up to you @gante to decide :-) 

Overall, great work! Thanks a lot @gmftbyGMFTBY for adapting the code so quickly here! 

**EDIT:** Sorry, actually let's please remove the deprecated `max_length` before merging - that's actually a "must-do" before merging IMO (so not a fully approval here :sweat_smile: )

gmftbyGMFTBY commented:
Okay, I am working on it! Thanks a lot for your reviews!

patrickvonplaten commented:
BTW @gmftbyGMFTBY,

Just read a through your extremely nice issue! It seems like you experimented with OPT as well, so maybe let's add a test for OPT as well then ? :-) OPT's `past_key_values` are slightly different compared to GPT2's `past_key_values` so maybe instead of adding a test for GPT-J and GPT-2, it would make more sense to add a test for OPT in addition to GPT2?

Also, if the paper is only concerned with open-ended generation (so less with encoder-decoder architectures), I'm also totally fine with **not** testing for T5 and BART (it's a nice to have, but if it takes too much time and it's not too important - happy to skip it!).

Regarding the fast dummy test, could you maybe make use of those dummy models: 
- https://huggingface.co/hf-internal-testing/tiny-random-gpt2
- https://huggingface.co/hf-internal-testing/tiny-random-gptj
- https://huggingface.co/hf-internal-testing/tiny-random-t5
- https://huggingface.co/hf-internal-testing/tiny-random-bart

The tests colud look very similar to:
https://github.com/huggingface/transformers/blob/71ca79448cd334970fa2893f4faaa094ca13ca6f/tests/generation/test_generation_utils.py#L2053

just much shorter, *i.e.* they only need to test for shape equality.

gmftbyGMFTBY commented:
Yeah, we have already tested the OPT models, and it works fine. I will supply more tests to the pre-trained models that you mentioned.

gmftbyGMFTBY committed:
Merge branch 'huggingface:main' into csearch-pr-v2

gmftbyGMFTBY committed:
add more test

gmftbyGMFTBY commented:
@patrickvonplaten more tests about these models are added: 
* gpt2-large
* gpt-j (EleutherAI/gpt-j-6B)
* opt (facebook/opt-6.7b)
* BART (facebook/bart-large-cnn)
* T5 (flax-community/t5-base-cnn-dm)

Thes tests are passed successfully. Can you do the final check about this PR?

gmftbyGMFTBY committed:
do the style, quality, consistency checks

gante commented:
Thank you for being part of this process @gmftbyGMFTBY 🙌 All queries have been addressed and the PR looks in a good state, merging!

yxuansu commented:
@gante @patrickvonplaten @sgugger -- Many thanks for your kind help throughout the process! It means a great deal to me and @gmftbyGMFTBY. Huggingface is the best!
